### ISSUE TITLE
Add automated component testing

### ISSUE THREAD

**Issue by ptgott** [gravitational/docs#153]
We recently (#136) added end-to-end browser tests to the docs using Playwright. We should also add tests for individual components, ideally without having to visit individual pages in `content/<ver>` since these change frequently.

One approach to this is to use Storybook alongside the Storybook [Test runner](https://storybook.js.org/docs/react/writing-tests/test-runner), which allows us to turn Storybook stories into Playwright tests. 

Let's consider this issue complete when:

- We have written at least one automated component test (e.g., for fixing a bug)
- We have included a command to run the automated component test in `package.json` and our CI config

**Comment by ptgott** [comment]
@hatched @avatus @sandylcruz Is there any other context we should add here? Any other approaches that would work better?

**Comment by hatched** [comment]
Having tests for the individual components is a good idea however I try to only reach for something like Playwright when I need to write an end to end test. Something in the context of the "bigger picture" like the whole app running connected to a real backend. 

I feel using a headless browser test runner like playwright for individual component tests is quite heavy weight compared to using something like jsom with Jest and [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/) while simulating interactions.

**Comment by ptgott** [comment]
@hatched That makes sense to me! I had been leaning toward Storybook because we have a bunch of `.stories` files sitting a previous commit in our old corpsite repo before we split `gravitational/docs` into its own repo. Those files don't exist in `gravitational/docs`, though, so it's a good opportunity to start afresh with lightweight component testing.

**Comment by ninaTorgunakova** [comment]
It would be great to implement automated component testing, I also want to suggest a solution with not only interaction tests for docs components:

Mostly, the components in the docs repo are really simple and don't contain hard logic connected with interactions. It will be handy to add such tests for the complex components (which are few), but I anticipate that even more useful for docs would be to have visual tests. They give the ability to catch regressions in UI, for example, if we change something in the code and unintentionally worse UI in unexpected places, visual tests will warn us about this. It will be especially helpful also because it's hard to every time check the visual regressions for all pages when there are a lot of them.

So, I propose the flow of implementing interaction tests and visual tests. It can be done by a single Storybook, which has [Test Runner](https://storybook.js.org/docs/react/writing-tests/test-runner) (which is powered by Jest by the way) for automating test executions and [Chromatic](https://storybook.js.org/docs/react/writing-tests/visual-testing) for visual tests. Both UI and interaction tests can be easily run automatically.

I can create a PR by implementing these two tools and creating a few stories for the most complex stories, and then we can decide if it covers all needs. WDYT?

**Comment by ninaTorgunakova** [comment]
@avatus Thanks, Michael, I will take a look at this PR.
Regarding Chromatic, I think it depends on your decision if this tool will be valuable for you for permanent use, did you have problems with visual regression before or not, and if there is a high probability of frequent changes in common files which can invoke such regressions. Seems to me, it can be valuable in this case with a lot of pages, and simple components. By heaviness, do you mean [its bundle size](https://bundlephobia.com/package/chromatic@6.18.0)?

**Comment by avatus** [comment]
By heavy i mean, cloud hosted with project tokens and all that related configuration. I think storybook alone would get us mostly what we need.

**Comment by ptgott** [comment]
@ninaTorgunakova @avatus Thanks for looking into this! I'd rather not tie us to another third-party SaaS vendor unnecessarily. We tend not to have issues with visual regression (as far as I know)—what I'm worried about with regard to our components are the more complex ones like `Tabs`, where it's easier to trigger a regression in one code path by fixing a bug in another.

I think Storybook sounds good!

Is there a more dependency-free alternative to Chromatic for snapshot testing if we just want a quick way to prevent unintended regressions in simple components? E.g., something like a golden file test where we render each React component, serialize it, and compare it to an on-disk serialization?

**Comment by ninaTorgunakova** [comment]
@ptgott @avatus Thanks for describing your opinion, I understood it.

If you don't have issues with visual regressions and are open to Storybook, then we can go with simple [Snapshot tests](https://storybook.js.org/docs/react/writing-tests/snapshot-testing) which are provided by Storybook. It's a very quick way to ensure that the DOM structure doesn’t change and that there are no errors appeared. They are easy to use, with no additional configurations/environments. Will this option be more suitable?

And sure, besides that, let's add interaction tests for complex components.

**Comment by ptgott** [comment]
@ninaTorgunakova Storybook snapshot tests sound good to me! This will be a quick way to ensure we have solid test coverage without needing to define test expectations retroactively for all of our components. Do you have any thoughts on this, @avatus?

**Comment by avatus** [comment]
Yup, sounds good to me!

**Comment by iAdramelk** [comment]
@avatus @ptgott Sorry for stepping in, but I wanted to share some considerations about snapshots and screenshots testing there.

We tried to use snapshots on four different projects, and on each of them, we removed them in the end. Reasons were:

- They are very noisy - every small change can potentially cause multiple snapshots to break.
- Their body is big chunks of automatically generated code. Comparing them visually takes a lot of time, and it is easy to miss something.
- They only show HTML code, but not the applied CSS and visual render. As a result, they can't catch visual regressions of problems with JS logic.

As a result, after some time, everybody just starts to automatically update them without reading.

On projects where we removed snapshots, we were usually switching to screenshots testing and it was a lot more effective. Comparing images are faster and easier than comparing snapshots and this comparison often helped in finding bugs.

Speaking of visual regressions on the docs site, I remember multiple times then I changed some base component like buttons, dropdown, or menu item styles and missed manually checking it on one of the pages, with some combination of content or in the mobile and broken version came to prod. I usually catch and fix these bugs myself, then retest in on prod, but having screenshot testing would help to catch these bugs before publishing or at least to add new tests to prevent this situation in the future.

Saying that, setting up screenshot testing manually takes a lot of time, resources, and infrastructure - to store old versions of screenshots, to compare them, to generate diffs, to update changed screenshots, etc. So usually, instead of doing it ourselves, we use third-party services like Chromatic.

We proposed Chromatic in this case because it is a service from the creators of Storybook, and it works with storybook pages out of the box - it can automatically create screenshots for each storybook page on deployment and automatically compare them on new branch creation.

I do understand that having additional SaaS services is not optimal, especially if docs styles are not changed very often, but having screenshot testing, in my opinion, is a lot more helpful than snapshots.

And even if we decide not to use Chromatic, I'd say let's still not do snapshot testing and only do unit tests. In my memory, I do not remember any cases then snapshots helped to catch a bug, but they always make tests more fragile and create a lot of noise.

**Comment by avatus** [comment]
@iAdramelk I love your input, thank you!!

I appreciate the feedback on storybook snapshot tests. Honestly, when it comes down to it, I don't mind how we do our testing and is mostly up to the docs team. Anything that can be automated by our writers/engineers to help keep our components in line is what I care about. I 👍 anything that achieves that goal. My preference is as little added complexity as possible but, you tell me.

**Comment by ptgott** [comment]
That's a really helpful writeup, @iAdramelk, thanks! @avatus @ninaTorgunakova Why don't we start with unit tests for complex components? I think setting up the SaaS dependency will require some non-dev work that we'll have to talk about within Teleport (procurement, etc.).

**Comment by ninaTorgunakova** [comment]
@ptgott Yes, I'm already in the process of adding unit tests and stories (started with `Tabs`). Ping me please when there is any news regarding the discussion of visual tests 👍

**Comment by ninaTorgunakova** [comment]
@ptgott PR with automated component testing was merged, but there is left our recommendation to add visual tests. If it's more convenient to track a new issue, it can be created as you proceed with agreements regarding Chromatic within Teleport (or just continue in this issue).

**Comment by ptgott** [comment]
Hi @ninaTorgunakova, I would say that the automated component testing PR fulfills most/all of the requirements for this issue, since we can now test complex behavior in our components. 

Preventing regressions is also important, but I think we can decouple that work from this task. Would you be okay creating an issue for visual tests? Then I can run the issue by some people at Teleport.

Does that make sense to you, @avatus ?

**Comment by ninaTorgunakova** [comment]
@ptgott Yes, sure, I added a separate issue for it with a description https://github.com/gravitational/docs/issues/328.

**Comment by ptgott** [comment]
Closing as completed in #312
